File: 6_tree/层序遍历.ts
/* eslint-disable prefer-destructuring */
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val
    this.left = left === undefined ? null : left
    this.right = right === undefined ? null : right
  }
}

function levelOrder(root: TreeNode | null): number[][] {
  if (!root) return []
  const queue: TreeNode[] = [root]
  const res: number[][] = []

  while (queue.length) {
    const level: number[] = []
    const step = queue.length
    for (let _ = 0; _ < step; _++) {
      const cur = queue.shift()! // eslint-disable-line @typescript-eslint/no-non-null-assertion
      level.push(cur.val)
      cur.left && queue.push(cur.left)
      cur.right && queue.push(cur.right)
    }
    res.push(level)
  }

  return res
}

export {}

File: 6_tree/层序遍历.py
# Definition for a binary tree node.
from collections import deque
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        """层序遍历"""
        if not root:
            return []

        res = []
        queue = deque([root])
        while queue:
            level = []
            len_ = len(queue)
            for _ in range(len_):
                cur = queue.popleft()
                level.append(cur.val)
                if cur.left:
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)
            res.append(level)
        return res
